### ISSUE TITLE
Children exclusion patterns

### ISSUE THREAD

**Issue by cassidydotdk** [kamsar/Unicorn#86]
So given this configuration snippet:

    <include database="master" path="/sitecore/content/Shared Data" name="Dev Shared Data">
             <exclude database="master" path="/sitecore/content/Shared Data/Contact Companies/" />
             <exclude database="master" path="/sitecore/content/Shared Data/Contacts/" />
     </include>

Idea being, include "Shared Data" and everything below except where otherwise specified. So to put a skeleton structure in place, for supporting "Default Locations" on a template field for instance, but leaving alone any content items in the tree.

Unicorn right now doesn't allow this. I believe it right now insists, to put Contacts under Unicorn control, and it's children too.

(certainly this is what the Content Editor ribbon is telling me)

**Comment by kamsar** [comment]
What do you think of making the child exclusion a bit less of an accident, something like:

`<include database="master" path="/sitecore/templates" children="false" />` for roots to include the single item or

```
<include database="master" path="/sitecore/templates">
     <exclude path="/sitecore/templates/lol" childrenOnly="true" />
</include>
```

To explicitly declare an exclude as being "for children of the path only"

**Comment by cassidydotdk** [comment]
I agree that having children included based solely on a trailing / is probably not ideal. The attribute approach you propose is better.

But I think a challenge is being overlooked. Say I wanted to include a root folder of some sort, like as for a default location for a datasource on a rendering.

    <include database="master" path="/sitecore/content/shared data/accordions" includeChildren="false">

But then, I would also need a few sample accordions deployed. If I added another <include> for these, would I not get slapped with the overlapping configuration exception?

**Comment by kamsar** [comment]
In my projects I have a special tree dedicated for sample content that gets around that issue (also makes it handy to write smoke tests against all sample pages during CI deploys - a trick that's saved my rear a few times).

When designing Rainbow I made the compromise that the tree format would be 'solid' so you cannot have a tree that includes children where the parents are not included. This was for performance as well as technical reasons (e.g. in some cases I cannot verify the correctness of the tree without knowing the ID of the parent item - not just the path, and if the parent item is not there, well...incorrectness can creep in).

Given that trees must be solid I made the choice to not support 'include then exclude then include' because that's not possible in all but the single case where you include a parent but only certain children.

Maybe we could do something like:
```
<include path="..." includeChildren="false">
    <include child="child" />
</include>
```

Seems like that would support your use case as well as the limitations of SFS.

**Comment by kamsar** [comment]
This has been implemented. I went with a slightly different grammar than we'd been talking about above: https://github.com/kamsar/Unicorn/blob/ce1e7aacdb65d53cffa617f8a11106dd70b944e0/src/Unicorn.Tests/Predicates/TestConfiguration.xml

This will be part of Unicorn 3.1, and can be perused right now in Unicorn 3.1-beta5 on NuGet

**Comment by kamsar** [comment]
Also not mentioned in the https://github.com/kamsar/Unicorn/commit/ce1e7aacdb65d53cffa617f8a11106dd70b944e0 commit message, exclude paths may now be relative to the include path by omitting the / in front of the exclude path. Allows things to be shorter and cleaner.

**Comment by cassidydotdk** [comment]
Oops, the email updates on this issue ended up deep in some obscure folder in my Gmail setup :-)  Looks pretty nifty indeed. Will get our current solution updated and see how it all fares.

**Comment by kamsar** [comment]
This is fixed in 3.1 final.
